Plugins/SwiftProtobufPlugin/plugin.swift

```diff
@@ -0,0 +1,139 @@
+
+        // We need to find the configuration file at the root of the target
+        guard let configurationFilePath = target.sourceFiles.first(where: { $0.path.lastComponent == Self.configurationFileName }) else {
+            throw PluginError.noConfigFile
+        }
+
+        let data = try Data(contentsOf: URL(fileURLWithPath: "\(configurationFilePath.path)"))
+        let configuration = try JSONDecoder().decode(Configuration.self, from: data)
+
+        // We need to find the path of protoc and protoc-gen-swift
+        let protocPath: Path
+        if let configuredProtocPath = configuration.protocPath {
+            protocPath = Path(configuredProtocPath)
+        } else {
+            protocPath = try context.tool(named: "protoc").path
+        }
+        let protocGenSwiftPath = try context.tool(named: "protoc-gen-swift").path
+
+        // This plugin generates its output into GeneratedSources
+        let outputDirectory = context.pluginWorkDirectory
+
+        return configuration.invocations.map { invocation in
+            self.invokeProtoc(
+                target: target,
+                invocation: invocation,
+                protocPath: protocPath,
+                protocGenSwiftPath: protocGenSwiftPath,
+                outputDirectory: outputDirectory
+            )
+        }
+    }
+
+    /// Invokes `protoc` with the given inputs
+    ///
+    /// - Parameters:
+    ///   - target: The plugin's target.
+    ///   - invocation: The `protoc` invocation.
+    ///   - protocPath: The path to the `protoc` binary.
+    ///   - protocGenSwiftPath: The path to the `protoc-gen-swift` binary.
+    ///   - outputDirectory: The output directory for the generated files.
+    /// - Returns: The build command.
+    private func invokeProtoc(
+        target: Target,
+        invocation: Configuration.Invocation,
+        protocPath: Path,
+        protocGenSwiftPath: Path,
+        outputDirectory: Path
+    ) -> Command {
+        // Construct the `protoc` arguments.
+        var protocArgs = [
+            "--plugin=protoc-gen-swift=\(protocGenSwiftPath)",
+            "--swift_out=\(outputDirectory)",
+            // We include the target directory as a proto search path
+            "-I",
+            "\(target.directory)",
+        ]
+
+        // Add the visibility if it was set
+        if let visibility = invocation.visibility {
+            protocArgs.append("--swift_opt=Visibility=\(visibility.rawValue.capitalized)")
+        }
+
+        var inputFiles = [Path]()
+        var outputFiles = [Path]()
+
+        for file in invocation.protoFiles {
+            // Append the file to the protoc args so that it is used for generating
+            protocArgs.append("\(file)")
+            inputFiles.append(target.directory.appending(file))
+
+            // The name of the output file is based on the name of the input file.
+            let protobufOutputName = file.replacingOccurrences(of: ".proto", with: ".pb.swift")
```

Comment:
This could be a little dangerous since it will replaces all locations of ".proto" in the path, I'd suggest doing a more targeted extension swap instead. Someone, somewhere will have "Source/Target/Support.protos/foo.proto"…

And should we validate that the file does end in .proto in the first case?  i.e. - catch another form of user error?

Reply:
Good points! I changed it to only swap the suffix and added some validation for the input file extensions.